### PR TITLE
Add internal solution for Circular Barn Revisited (usaco622)

### PR DESCRIPTION
**Pull request by ChuangSheep**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/usaco-622.mdx [hunk 1]
```diff
@@ -0,0 +1,81 @@
+---
+id: usaco-622
+source: USACO Gold 2016 February
+title: Circular Barn Revisited
+author: Chuyang Wang
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_cbarn2_gold_feb16.html)
```
Comment (SansPapyrus683) [inline solutions/usaco-622.mdx:8]:
```suggestion
[Official Analysis (C++)](http://www.usaco.org/current/data/sol_cbarn2_gold_feb16.html)
```

File: solutions/usaco-622.mdx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+## Explanation
+
+The problem asks for the minimum amount of distance the cows need to travel for a circular barn. However, if we fix the position of the first door, the problem can be reduced to a linear barn $\texttt{rooms}[1..n]$ where we place the first door at the beginning of our linear barn. In such case, we can define $\texttt{dp}[i][j]$ as the sum of distance to fill the rooms $[j, n]$ if we place the $i$-th door at $j$. All values of the $\texttt{dp}$ array are initialized with infinity. We further set $\texttt{dp}[0][n+1] = 0$ because with no doors the only case where the distance is zero is to fill no room. Our transitions would then be
+
+$\texttt{dp}[i][j] = min(\texttt{dp}[i][j], \sum_{p=j+1}^{n+1} (p-j-1) \cdot \texttt{rooms}[p-1] + \texttt{dp}[i-1][p])$,
```
Comment (SansPapyrus683) [inline solutions/usaco-622.mdx:14]: what does this mean in normalspeak

File: solutions/usaco-622.mdx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	freopen("cbarn2.out", "w", stdout);
+
+	int n, k;
+	cin >> n >> k;
+	deque<int> rooms(n);
+	for (int &r : rooms) {
+		cin >> r;
+	}
+
+	long long best_answer = INF;
```
Comment (SansPapyrus683) [inline solutions/usaco-622.mdx:42]:
```suggestion
	long long best_answer = INT64_MAX;
```

File: solutions/usaco-622.mdx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+There are $n$ doors which can be chosen as the first door. In each DP, we add each time one of the $k$ doors and iterate through each of $n$ positions to place it. For each of these positions, we go through all possible $n$ placement of the last door and calculate the optimal result for our current position with the new door. This yields a total time complexity of $\mathcal{O}(k n^3)$, which is fast enough for the given constraints. 
+
+## Implementation
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const long long INF = 1e14;
```
Comment (SansPapyrus683) [inline solutions/usaco-622.mdx:29]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-622.mdx (hunk 3 above)
Comment (ChuangSheep) [reply solutions/usaco-622.mdx:42]: we had this discusstion before...  I don't think this is the best way but ok ima gonna change it here too

File: solutions/usaco-622.mdx (hunk 2 above)
Comment (ChuangSheep) [reply solutions/usaco-622.mdx:14]: added an explanation for the equation after it

File: solutions/usaco-622.mdx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	freopen("cbarn2.out", "w", stdout);
+
+	int n, k;
+	cin >> n >> k;
+	deque<int> rooms(n);
+	for (int &r : rooms) {
+		cin >> r;
+	}
+
+	long long best_answer = INT64_MAX;
```
Comment (SansPapyrus683) [inline solutions/usaco-622.mdx:41]: rename `best_answer` to smth better
`min_dist`? if i remember how this problem works

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/usaco-622.mdx [hunk 6]
```diff
@@ -67,7 +67,7 @@ int main() {
 			}
 		}
 		// Update the best answer using the current dp answer
-		best_answer = min(best_answer, dp[k][0]);
+		min_dist = min(min_dist, dp[k][0]);
```
Comment (SansPapyrus683) [inline solutions/usaco-622.mdx:70]: u sure `min_dist` accurately represents the answer?

Comment (ChuangSheep) [reply solutions/usaco-622.mdx:70]: the answer is the minimum amount of distance that all cows need to travel to fill the rooms, so I think `min_dist` is alright.

**Comment by SansPapyrus683** [comment]
@envyaims

**Review by envyaims — APPROVED** [review]
woah

**Pull request merged**